redis/tests/test_cache.rs

```diff
@@ -0,0 +1,438 @@
+                .query_async(&mut con)
+                .await?;
+            assert_hit(&con, 0);
+            assert_miss(&con, 2);
+
+            let _: Vec<Option<String>> = redis::cmd("MGET")
+                .arg("key_1")
+                .arg("key_2")
+                .query_async(&mut con)
+                .await?;
+            assert_hit(&con, 2);
+            assert_miss(&con, 2);
+
+            let _: Option<String> = redis::cmd("GET").arg("key_1").query_async(&mut con).await?;
+            assert_hit(&con, 3);
+            assert_miss(&con, 2);
+            Ok::<_, RedisError>(())
+        },
+        runtime,
+    )
+    .unwrap();
+}
+
+#[rstest]
+#[case::tokio(RuntimeType::Tokio)]
+#[cfg_attr(feature = "async-std-comp", case::async_std(RuntimeType::AsyncStd))]
+fn test_cache_is_not_target_type_dependent(#[case] runtime: RuntimeType) {
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(
+        async move {
+            let mut con = ctx.async_connection_with_cache().await?;
+            let _: () = con.set("KEY", "77").await?;
+            let x: u32 = con.get("KEY").await?;
+            assert_eq!(x, 77);
+            let x: String = con.get("KEY").await?;
+            assert_eq!(x, "77");
+            let x: u8 = con.get("KEY").await?;
+            assert_eq!(x, 77);
+            assert_hit(&con, 2);
+            assert_miss(&con, 1);
+            Ok::<_, RedisError>(())
+        },
+        runtime,
+    )
+    .unwrap();
+}
+
+#[rstest]
+#[case::tokio(RuntimeType::Tokio)]
+#[cfg_attr(feature = "async-std-comp", case::async_std(RuntimeType::AsyncStd))]
+fn test_cache_with_pipeline(#[case] runtime: RuntimeType, #[values(true, false)] atomic: bool) {
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+    block_on_all(
+        async move {
+            let mut con = ctx.async_connection_with_cache().await?;
+            // Test cache for both atomic and non-atomic Pipeline and mix MGET,GET,ignore in the pipeline.
+            let ((k1, k2),): ((i32, i32),) = get_pipe(atomic)
+                .cmd("SET")
+                .arg("key_1")
+                .arg(42)
+                .ignore()
+                .cmd("SET")
+                .arg("key_2")
+                .arg(43)
+                .ignore()
+                .cmd("MGET")
+                .arg(&["key_1", "key_2"])
+                .query_async(&mut con)
+                .await?;
+
+            assert_eq!(k1, 42);
+            assert_eq!(k2, 43);
+            // There are 2 miss for key_1, key_2 used with MGET
+            assert_hit(&con, 0);
+            assert_miss(&con, 2);
+
+            let ((k1, k2), k3, k4): ((i32, i32), i32, Option<i32>) = get_pipe(atomic)
+                .cmd("MGET")
+                .arg(&["key_1", "key_2"])
+                .cmd("GET")
+                .arg("key_1")
+                .cmd("GET")
```

Comment:
can this be moved one command earlier, just to verify that the mechanism isn't order-dependent?

Reply:
I've put MGET in between them